### PR TITLE
esp8266: uart add missing poll support feature

### PR DESCRIPTION
**Pull request by mhoffma**
adding poll feature for review.

### CODE REVIEW

**Comment by pfalcon** [comment]
Thanks. Please follow commit message conventions described in https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md .

**Comment by mhoffma** [comment]
Ok so I used the prefix what is missing? Do you want me to remove uart, add a colon or something else?

**Comment by mhoffma** [comment]
is this a better comment?

File: esp8266/uart.c [hunk 1]
```diff
@@ -200,6 +200,19 @@ bool uart_rx_wait(uint32_t timeout_us) {
...
 }
 
+int uart_rx_any(uint8 uart) {
+    if (input_buf.iget != input_buf.iput) {
+        return true; // have at least 1 char ready for reading
+    }
+    return false;
+}
+int uart_tx_any(uint8 uart) {
+    uint32 fifo_cnt = READ_PERI_REG(UART_STATUS(uart)) & (UART_TXFIFO_CNT<<UART_TXFIFO_CNT_S);
```
Comment (dpgeorge) [inline esp8266/uart.c]: Spaces around the "<<".

File: esp8266/uart.c [hunk 2]
```diff
@@ -200,6 +200,19 @@ bool uart_rx_wait(uint32_t timeout_us) {
...
+    if ((fifo_cnt >> UART_TXFIFO_CNT_S & UART_TXFIFO_CNT) == 0)
+      return false;
```
Comment (dpgeorge) [inline esp8266/uart.c]: Please use curly braces around this single line (follow existing conventions).

Comment (mhoffma) [inline esp8266/uart.c]: ok, done.

File: esp8266/uart.c [hunk 3]
```diff
@@ -200,6 +200,20 @@ bool uart_rx_wait(uint32_t timeout_us) {
     }
 }
 
+int uart_rx_any(uint8 uart) {
+    if (input_buf.iget != input_buf.iput) {
+        return true; // have at least 1 char ready for reading
+    }
+    return false;
+}
+int uart_tx_any(uint8 uart) {
```
Comment (dpgeorge) [inline esp8266/uart.c]: I think the logic/naming/use of this function is wrong.  It returns true if there are characters waiting to be transmitted.  Instead we want a function that tells us if there is room in the tx buffer.

**Comment by dpgeorge** [comment]
The logic for the newly named uart_tx_any_room function was still inverted, but I fixed it, squashed the changes and merged in 91eb0153d30420618b06efd18631490ac7fbfaae.

**Comment by pfalcon** [comment]
What about testing - was it tested to work as expected (under "load"), or do we just assume that it works? ;-)

**Comment by dpgeorge** [comment]
> What about testing

I tested it as far as I could: polling for rx works as expected (blocks until chars come in), and polling for tx always returns "room available" even after sending a huge amount of data out.  That's probably because the uart tx function always waits for the output buffer to have enough room before sending.

**Comment by mhoffma** [comment]
I tested as in used it. So does that means we could flood it and never hit
that condition as it stands right now?

**Comment by peterhinch** [comment]
If an asynchronous application allocates a buffer and writes it to the UART I'd hope that tx polling would block until the buffer was sent or a timeout occurred. Synchronous code blocks in this way.

There needs to be some way to avoid trying to send data faster than the interface allows: in its absence the code might allocate ever more buffers or overwrite a preallocated buffer.

**Pull request closed without merging**